### PR TITLE
feat(flexslider): no opacity anymore

### LINKED ISSUE #30893: Single Image in product gallery generate CLS with FlexSlider
**Prerequisites (mark completed items with an [x]):**
- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
Right now if you have a product with a gallery with FlexSlider you have a CLS of 0.13~ just because of that image on mobile. This is enough to alert Google Lighthouse, that has a max of 0,1 to avoid notices (as we are above there is an alert).
Some examples documented on the web about the issue:

* https://www.commercegurus.com/woocommerce-product-gallery-speed/ - Check "Opacity transition on Product Gallery markup wrapper.", removing the css inline and hiding them magically fix but if you use visibility or display:none to not show it (to do it later) the image is missing, I guess that something on FlexSlider doesn't work if the div is not visible (I checked the code and there are various check for visible)
* https://wordpress.org/support/topic/single-product-images-cause-cumulative-layout-shift/ - another user did with visibility and executing manually flexslider to do the same trick and show the gallery at the end
* https://github.com/woocommerce/woocommerce/pull/30701 - This is another error from LightHouse but the PR fix

Honestly it is surprising that WooCommerce has this issues with all the users around the world and how much the Core Web Vitals are important for all the users and customers of Automattic. 

**Expected behavior**
Don't have CLS at all.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Go to single product with Chrome
2. Check the CLS

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by Mte90**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Remove the opacity to the gallery that is reported as CLS from Lighthouse and is an old way to hide the gallery lazyload. Now WP and browser have a native way for images and a lot of things are happened in the last 6 years. Also after this https://github.com/woocommerce/woocommerce/pull/31407 some information are now in the page to improve the layout and size of the gallery.
The only doubts is that some CSS was inline in the templates folder, so as is removed now for all the themes around that copied still have this CSS. With those changes the opacity is not removed, so it is required a workaround to flip that value as true if opacity is set. Before to do that I am looking for feedbacks.

Closes #30893.

This changes is by me and @erikyo as we talked about how to do it in the best way.

### How to test the changes in this Pull Request:

1. Go to single product with Chrome
2. Check the CLS

### FOR PR REVIEWER ONLY:

* [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/legacy/js/frontend/single-product.js [hunk 1]
```diff
@@ -93,7 +93,6 @@ jQuery( function( $ ) {
 
 		// No images? Abort.
 		if ( 0 === this.$images.length ) {
-			this.$target.css( 'opacity', 1 );
```
Comment (Mte90) [inline plugins/woocommerce/legacy/js/frontend/single-product.js:96]: Maybe we can move this line outside the if to let oldest template to work with this changes.

**Review by jeffstieler — CHANGES_REQUESTED** [review]
Thank you for your contribution @Mte90!

This looks good on products with only one image, but the opacity changes cause other layout shifts on products with multiple images (while images are loading). See this screen capture of a page refresh:

Capture 2022-01-18 at 09 57 43

**Comment by Mte90** [comment]
So probably hide all the images except the first one should fix that user case. What do you think @erikyo?
maybe with a css class to enqueue to all the images and that JS remove after loading?

**Comment by erikyo** [comment]
@jeffstieler maybe the [patch](https://github.com/woocommerce/woocommerce/blob/c9696e66b9ebe1adaf6bf8a0f28fe80f38dae220/plugins/woocommerce/legacy/js/flexslider/jquery.flexslider.js#L247) that set the with and the height of the thumbnail images isn't present in your test environment? In my blog this fix work as it should, check [this](https://modul-r.codekraft.it/product/logo-collection/)

After more extensive testing I believe I have figured out why.... the reason was the thumbnail images were loaded after the main image **slider** (we are displaying only one of the large image but after this the subsequent images were loaded (this is another fix to do, because many kb can be saved avoiding this). 

The glitch that is displayed by @jeffstieler is caused by nav thumbnails srcset that finds one of the images and uses it, even if it is much larger than the requested one. So it find the large image of the slider (600px X 600px) showing it until the smaller image was downloaded. Will find a solution with @Mte90, I guess that also thumbnail loading must be anticipated or an appropriate srcset must be used. 
image

**Comment by jeffstieler** [comment]
@erikyo looks like you've identified the root cause of the glitch. Thanks again for taking the time to make these fixes. (My local test environment did have the latest patch to `jquery.flexslider.js`, fwiw)

**Comment by erikyo** [comment]
@jeffstieler I'm sorry but what I said above is inaccurate and after further investigation seems the issue was a little more complex and related in how the slideshow and its nav are created... I try explain better with some screenshots the loading sequence:
image
1) before the flexslider is loaded ($( 'body' ).on( 'init'..) the scenario is this, the first image is displayed at 100% of the container width and the following slides at 25% of the width. There is not yet the "real" nav, in fact the first image is missing.
image
2) then the slider images are resized to 100% of the width
image
3) the images after being resized are aligned to the first one (actually this happens because Flexslider sets the width of the container)
image
4) finally the nav is created using the data taken from the images (that's why each div has the data-thumb attribute)
---
We are looking (with @Mte90) for a solution to make these things happen without upsetting flexslider, it would be enough for the images to be moved at the moment that the nav was created and only after this apply the resize to the hidden images (slider2, slider3, slider4,...).

**Comment by Mte90** [comment]
This last changes should improve the situation, @erikyo says that now doesn't load anymore 3 images at full size so LightHouse is very happy of that (and the browsers too).
Looking at the code is missing to do some changes to the fade part.

**Comment by erikyo** [comment]
These should also fix the other types of slideshows that are not being used... from my tests both were not working before, is it possible? @Mte90 @jeffstieler 

To perform a quick test, just use the filter:
```
add_filter( 'woocommerce_single_product_carousel_options', 'filter_single_product_carousel_options' );
function filter_single_product_carousel_options( $options ) {
	$options['direction'] = 'vertical';
	// $options['animation'] = 'fade';
	return $options;
}
```

**Comment by erikyo** [comment]
This makes sure that the thumbnails that are set with the `woocommerce_product_thumbnails_columns` filter are displayed correctly. Moreover this commit will speed up loading because the width and the height of thumbnails are pre-calculated using the number of columns (which [was there](https://github.com/woocommerce/woocommerce/blob/b5ae232106cd5d0d13869a0b030f0f30e5cd3e53/plugins/woocommerce/templates/single-product/product-image.php#L39) but not used). 
Thanks to srcset under certain conditions are "reclycled" from the larger image and the browser may opt not to download the thumbnail using the sharper image. This has a huge benefit with lighthouse/mobile rating.

`add_filter( 'woocommerce_product_thumbnails_columns', function () { return 5; } ); // 3,4,5 are allowed values`

Those using non-square thumbnails might see the image not filling its box, but this could be fixed easily setting a fixed height to navigation items, with something like:
```
body.woocommerce div.product div.images .flex-control-thumbs li {
    height: 100px;
    margin: 10px auto;
}
.woocommerce div.product div.images .flex-control-thumbs li img {
    height: 100%;
    object-fit: contain;
    /* object-position: left; */
}
```

**Comment by Mte90** [comment]
There is any opportunity to get this change deployed? 
My customers are asking to improve the performances and I can't do anything as it is a Woocommerce fault and they don't understand that is not something we can do until this is merged (yes I can dequeue the script etc) but I can't do it for all my customers.

**Comment by peterfabian** [comment]
A follow-up PR is being created by @erikyo atm, so please don't review this one just yet. Thx!

**Comment by Mte90** [comment]
So I think that this can be closed for https://github.com/woocommerce/woocommerce/pull/33041

**Comment by peterfabian** [comment]
Thanks, closing.

**Pull request closed without merging**